### PR TITLE
[no release notes] PaxosLeaderElectionServiceBuilder

### PR DESCRIPTION
**Pull request by jeremyk-91**
**Goals (and why)**:

- Avoid having to use an error-prone 9 argument constructor with three longs that mean different things; see #1463.

**Implementation Description (bullets)**:

- Add `PaxosLeaderElectionServiceBuilder` that is a builder for `PaxosLeaderElectionService`.
- I'd rather not remove the 9-arg constructor (or decrease its visibility) because it is used by large internal product.

**Concerns (what feedback would you like?)**:

- I didn't add validation to the builder since the original constructor does not do this. Not sure if it should be added (currently you can have empty lists for learners, nulls and/or negative timeouts), but I went for the minimal change here.

**Where should we start reviewing?**: `PaxosLeaderElectionServiceBuilder`, then I guess double checking that I haven't done anything silly in switching Leaders or PaxosConsensusTestUtils over.

**Priority (whenever / two weeks / yesterday)**: whenever.

### CODE REVIEW

File: atlasdb-config/src/main/java/com/palantir/atlasdb/factory/Leaders.java [hunk 1]
```diff
@@ -116,16 +117,17 @@ public static LocalPaxosServices createLocalServices(
...
-                ImmutableList.copyOf(learners),
-                executor,
-                config.pingRateMs(),
-                config.randomWaitBeforeProposingLeadershipMs(),
-                config.leaderPingResponseWaitMs());
+        PaxosLeaderElectionService leader = new PaxosLeaderElectionServiceBuilder()
+                .proposer(proposer)
+                .knowledge(ourLearner)
+                .potentialLeadersToHosts(otherLeaders)
+                .acceptors(ImmutableList.copyOf(acceptors))
```
Comment (gsheasby) [inline atlasdb-config/src/main/java/com/palantir/atlasdb/factory/Leaders.java]: Should we have the builder take care of making an immutable copy?

File: leader-election-impl/src/main/java/com/palantir/leader/PaxosLeaderElectionServiceBuilder.java [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+                knowledge,
+                potentialLeadersToHosts,
+                acceptors,
+                learners,
+                executor,
+                pingRateMs,
+                randomWaitBeforeProposingLeadershipMs,
+                leaderPingResponseWaitMs);
+    }
+}
```
Comment (gsheasby) [inline leader-election-impl/src/main/java/com/palantir/leader/PaxosLeaderElectionServiceBuilder.java]: nit: newline

File: atlasdb-config/src/main/java/com/palantir/atlasdb/factory/Leaders.java (hunk 1 above)
Comment (jeremyk-91) [inline atlasdb-config/src/main/java/com/palantir/atlasdb/factory/Leaders.java]: Yep, I think so - changed.

File: leader-election-impl/src/main/java/com/palantir/leader/PaxosLeaderElectionServiceBuilder.java (hunk 2 above)
Comment (jeremyk-91) [inline leader-election-impl/src/main/java/com/palantir/leader/PaxosLeaderElectionServiceBuilder.java]: Alright, added.

File: docs/source/release_notes/release-notes.rst [hunk 3]
```diff
@@ -80,6 +80,11 @@ develop
     *    - |devbreak|
          - The persistent lock release endpoint has now been renamed to ``releaseBackupLock`` since it is currently only supposed to be used for the backup lock.
 
+    *    - |deprecated|
+         - ``PaxosLeaderElectionService`` is now deprecated, to mitigate risks of users supplying parameters in the wrong order.
```
Comment (gsheasby) [inline docs/source/release_notes/release-notes.rst]: This makes it sound like the whole service is deprecated - you mean "The public PLES constructor is now deprecated".

Comment (gsheasby) [inline docs/source/release_notes/release-notes.rst]: You mean that the constructor is deprecated - this makes it sound like the whole service has been deprecated.

Comment (jeremyk-91) [inline docs/source/release_notes/release-notes.rst]: Heh, good point. Yep, will fix this

**Pull request merged**